### PR TITLE
feat: signature via remote server

### PR DESCRIPTION
**Pull request by RaitoBezarius**
This PR is composed of ~four changes:

- generalization of signature schemes in the tool
- generalization of how to assemble stubs in the shared library
- a remote signature server that sends you signed stub in exchange for signature requests
- an implementation inside lzbt-systemd of remote signing

This enables moving the local signature to another server… or even maybe a hardware security token!

In the future, the server can ask for attestations or any kind of things to give the signed stub.

### CODE REVIEW

File: rust/tool/server/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+behind a reverse proxy with authentication or authorization or in a trusted network via a VPN.
+
+No rate-limit is applied to protect against denial of service, PRs are welcome to figure out a reasonable solution on that,
+otherwise rate-limits can be applied at the system level.
+
+## Endpoints
+
+- `POST /sign-stub`: assembles a signed stub based on the stub parameters sent, 200 OK with a signed binary as body, 400 with a plaintext error if failed.
+- `POST /sign-store-path`: assembles a signed binary based on the store path sent, 200 OK with a signed binary as body, 400 with a plaintext error if failed.
+- `GET /verify`: verify that the binary sent is signed according to the current keyring, returns a JSON `{ signed: bool, valid_according_to_secureboot_policy: bool }`, a signed binary can be invalid for the current Secure Boot policy, the two attributes represents this fact.
```
Comment (cole-h) [inline rust/tool/server/README.md:29]:
```suggestion
- `POST /verify`: verify that the binary sent is signed according to the current keyring, returns a JSON `{ signed: bool, valid_according_to_secureboot_policy: bool }`, a signed binary can be invalid for the current Secure Boot policy, the two attributes represents this fact.
```

Comment (RaitoBezarius) [reply rust/tool/server/README.md:29]: thx you madman

**Comment by blitz** [comment]
I would need some high-level information about the security model here. If people still need to protect the credentials to access this server, which signs everything, why not just given them the private keys themselves? The only advantage I can see now is that the server can log what it signed.

**Comment by RaitoBezarius** [comment]
Let's discuss it tomorrow :)

Le dim. 15 oct. 2023, 14:57, Julian Stecklina ***@***.***> a
écrit :

**Comment by RaitoBezarius** [comment]
This will have to wait until I have some time to fix the merge conflicts. Closing in the meantime, **do not delete the branch**.

**Pull request closed without merging**